packages/itwin/grouping-mapping-widget/src/widget/components/QueryBuilder.ts

```diff
@@ -536,122 +531,78 @@ export class QueryBuilder {
-                  queryString += ` AND ${joinClassName}.${property.name}='${property.value}'`;
-                } else {
-                  if (this._isFloat(property.value)) {
-                    queryString += ` AND ROUND(${joinClassName}.${property.name}, `;
-                    queryString += `${QueryBuilder.DEFAULT_DOUBLE_PRECISION})=`;
-                    queryString += `${Number(property.value).toFixed(
-                      QueryBuilder.DEFAULT_DOUBLE_PRECISION,
-                    )}`;
-                  } else {
-                    queryString += ` AND ${joinClassName}.${property.name}=${property.value}`;
-                  }
-                }
-              }
-            }
-          } else {
-            queryString += ` JOIN ${joinClassName} ON ${joinIdName} = ${baseIdName}`;
-            for (const property of joinClass.properties) {
-              if (property.isCategory) {
-                queryString += this.categoryQuery(property.value.toString());
-              } else {
-                if (property.needsQuote) {
-                  queryString += ` AND ${joinClassName}.${property.name}='${property.value}'`;
-                } else {
-                  if (this._isFloat(property.value)) {
-                    queryString += ` AND ROUND(${joinClassName}.${property.name}, `;
-                    queryString += `${QueryBuilder.DEFAULT_DOUBLE_PRECISION})=`;
-                    queryString += `${Number(property.value).toFixed(
-                      QueryBuilder.DEFAULT_DOUBLE_PRECISION,
-                    )}`;
-                  } else {
-                    queryString += ` AND ${joinClassName}.${property.name}=${property.value}`;
-                  }
-                }
-              }
-            }
+          const joinSegment = joinClass.isRelational
+            ? ` JOIN ${joinClassName}`
+            : ` JOIN ${joinClassName} ON ${joinIdName} = ${baseIdName}`;
+          querySegments.push(joinSegment);
+
+          for (let j = 0; j < joinClass.properties.length; j++) {
+            const prefix = j === 0 && joinClass.isRelational ? "ON" : "AND";
+            const propertySegment = this._propertySegment(joinClassName, joinClass.properties[j], prefix);
+            querySegments.push(propertySegment);
+          }
+          // when base is regular property, link base to first joined relational property
+          if (joinClass.isRelational && !baseClass.isRelational && i === 1) {
+            querySegments.push(` AND ${joinIdName} = ${baseIdName}`);
           }
         }
       }
 
-      const properties = baseClass.properties;
-      if (properties.length > 0) {
-        if (properties[0].isCategory) {
-          queryString += this.categoryQuery(properties[0].value.toString());
+      const whereSegment = this._whereSegment(baseClass, baseClassName);
+      querySegments.push(whereSegment);
+
+      const unionSegment = " UNION ";
+      querySegments.push(unionSegment);
```

Comment:
Store each `SELECT` of the `UNION` in an array and then `.join(" UNION ")` them.

Reply:
Changed to storing `SELECT`s to union segment and joining with `.join(" UNION ")`